plugins/cake/cake_

```diff
@@ -0,0 +1,202 @@
+#!/bin/bash
+# -*- sh -*-
+
+: << =cut
+
+=pod
+
+=encoding UTF-8
+
+=head1 NAME
+
+cake_ - Plugin to monitor cake's backlog, dropped, overlimits and requeues
+
+=head1 CONFIGURATION
+
+None needed.
+
+=head1 INTERPRETATION
+
+Cake, also known as sch_cake is a modern bandwidth limiter, which eliminates
+buffer bloat over slow links. It's also capable to give flows, hosts and each 
+flow of each host a fair part of the avaible bandwidth.
+
+This plugin allows for a monitor of the pressure on the qdisc, by monitoring key
+values.
+
+=head1 SEE ALSO
+
+Take a look at "man cake" to get more information about cake.
+
+=head1 MAGIC MARKERS
+
+ #%# family=auto
+ #%# capabilities=autoconf suggest
+
+=head1 AUTHORS
+
+RubenKelevra <ruben@vfn-nrw.de>
+
+work based on the tc plugin, authors:
+Steve Schnepp <steve.schnepp@gmail.com>,
+Samuel Smith <esaym@cpan.org>,
+Nye Liu <nyet@nyet.org>
+
+=head1 LICENSE
+
+GPLv2 or later
+
+=cut
+
+DEVICE=${0##*/cake_}
+
+tc_get_ifb_dev() {
+	dev="$(/sbin/tc filter show dev $DEVICE parent ffff: protocol all | grep "mirred" | grep "Egress Redirect to device")"
+	[ $? -eq 1 ] && return 1
+	
+	echo "$dev" | sed -n 's/.* device \([^ ]*\).*/\1/p' | tr ')' ' ' | awk '{ print $1 }'
+}
+
+tc_cake_sent() {
+	/sbin/tc -s qdisc show dev "$1" | grep -E "^ Sent" | tr ',' ' ' | tr ')' ' ' 
+}
+tc_cake_backlog() {
+	/sbin/tc -s qdisc show dev "$1" | grep -E "^ backlog" | tr 'p' ' ' 
+}
+
+DEVICE_IN="$(tc_get_ifb_dev)"
+has_if_in=$?
+
+case "$1" in
+	autoconf)
+	if [ -r /proc/net/dev ]; then
```

Comment:
Is there really no specific test for the actual use of `cake` on the system?

Reply:
Not in all cases. You could check if the sch_cake module is loaded, but if it's compiled as a built-in module that won't catch it. Also, the module can be loaded without being active.

The only sure way to see if cake is active is to do a `tc qdisc` and loop through all interfaces to try to find it. You could probably do `tc -j qdisc` and construct a query with jq that will spit out all cake instances...

Reply:
How about this one?
```shell
tc -j qdisc | jq -r '.[] | select(.kind == "cake").dev' | sort | uniq
```

Reply:
This looks great, thanks!

Reply:
So... I'll leave it the way it it. If you want to replace this function, feel free to suggest one which covers all features.

Also make sure to add a simple test, if jq is installed with an error-output for autoconf - for example it isn't installed on my system - never had the need for it - so an output about this dependency would be nice.

Reply:
> But do we really want to get rid of the filter with packets received?

I am not sure, what this means. How would you detect this situation? I would be happy to suggest a suitable filter.

Regarding the check of for `jq` - this is simply part of "autoconf", too:
```shell
if which jq >/dev/null; then
    echo "no (missing 'jq' executable)"
elif [ -r /proc/net/dev ]; then
...
```

Reply:
OK - I see.
Maybe we have different perceptions regarding the purpose of the "autoconf" test.
Here the plugin needs to answer two questions:
* can this plugin run at all? (i.e. are all dependencies satisfied?)
* is this plugin useful for this host?

The "autoconf" call should return "yes" only if both questions are affirmed.

Your current implementation of "autoconf" only cares (partly) for the first question.
Additionally it should check two more details:
* is `tc` available? (see my example for `jq` above)
* is "cake" in use on this system?

The latter can probably be tested like this:
```shell
[ -n "$(tc -j qdisc | jq -r '.[] | select(.kind == "cake").dev')" ]
```